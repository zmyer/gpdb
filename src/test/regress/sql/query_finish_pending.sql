drop table if exists _tmp_table;
create table _tmp_table (i1 int, i2 int, i3 int, i4 int);
insert into _tmp_table select i, i % 100, i % 10000, i % 75 from generate_series(0,99999) i;

-- make sort to spill
set statement_mem="2MB";
set gp_enable_mk_sort=on;
set gp_cte_sharing=on;

\! gpfaultinjector -f execsort_mksort_mergeruns -y reset --seg_dbid 2
-- set QueryFinishPending=true in sort mergeruns. This will stop sort and set result_tape to NULL
\! gpfaultinjector -f execsort_mksort_mergeruns -y finish_pending --seg_dbid 2

-- return results although sort will be interrupted in one of the segments 
select DISTINCT S from (select row_number() over(partition by i1 order by i2) AS T, count(*) over (partition by i1) AS S from _tmp_table) AS TMP;

\! gpfaultinjector -f execsort_mksort_mergeruns -y status --seg_dbid 2

-- test if shared input scan deletes memory correctly when QueryFinishPending and its child has been eagerly freed,
-- where the child is a Sort node
drop table if exists testsisc;
create table testsisc (i1 int, i2 int, i3 int, i4 int); 
insert into testsisc select i, i % 1000, i % 100000, i % 75 from
(select generate_series(1, nsegments * 50000) as i from 
	(select count(*) as nsegments from gp_segment_configuration where role='p' and content >= 0) foo) bar; 

set gp_resqueue_print_operator_memory_limits=on;
set statement_mem='2MB';
set gp_enable_mk_sort=off;

\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
-- Set QueryFinishPending to true after SharedInputScan has retrieved the first tuple. 
-- This will eagerly free the memory context of shared input scan's child node.  
\! gpfaultinjector -f execshare_input_next -y finish_pending --seg_dbid 2

select COUNT(i2) over(partition by i1)
from testsisc
LIMIT 2;

\! gpfaultinjector -f execshare_input_next -y status --seg_dbid 2

-- test if shared input scan deletes memory correctly when QueryFinishPending and its child has been eagerly freed,
-- where the child is a Sort node and sort_mk algorithm is used

set gp_enable_mk_sort=on;

\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
-- Set QueryFinishPending to true after SharedInputScan has retrieved the first tuple. 
-- This will eagerly free the memory context of shared input scan's child node.  
\! gpfaultinjector -f execshare_input_next -y finish_pending --seg_dbid 2

select COUNT(i2) over(partition by i1)
from testsisc
LIMIT 2;

\! gpfaultinjector -f execshare_input_next -y status --seg_dbid 2

reset gp_enable_mk_sort;
-- Disable faultinjectors
\! gpfaultinjector -f execsort_mksort_mergeruns -y reset --seg_dbid 2
\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
