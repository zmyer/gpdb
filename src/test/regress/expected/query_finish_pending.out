drop table if exists _tmp_table;
NOTICE:  table "_tmp_table" does not exist, skipping
create table _tmp_table (i1 int, i2 int, i3 int, i4 int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i1' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
insert into _tmp_table select i, i % 100, i % 10000, i % 75 from generate_series(0,99999) i;
-- make sort to spill
set statement_mem="2MB";
set gp_enable_mk_sort=on;
set gp_cte_sharing=on;
\! gpfaultinjector -f execsort_mksort_mergeruns -y reset --seg_dbid 2
20170125:11:06:26:002993 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execsort_mksort_mergeruns -y reset --seg_dbid 2
20170125:11:06:26:002993 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:26:002993 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- set QueryFinishPending=true in sort mergeruns. This will stop sort and set result_tape to NULL
\! gpfaultinjector -f execsort_mksort_mergeruns -y finish_pending --seg_dbid 2
20170125:11:06:26:003005 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execsort_mksort_mergeruns -y finish_pending --seg_dbid 2
20170125:11:06:26:003005 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:26:003005 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- return results although sort will be interrupted in one of the segments 
select DISTINCT S from (select row_number() over(partition by i1 order by i2) AS T, count(*) over (partition by i1) AS S from _tmp_table) AS TMP;
 s 
---
 1
(1 row)

\! gpfaultinjector -f execsort_mksort_mergeruns -y status --seg_dbid 2
20170125:11:06:27:003020 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execsort_mksort_mergeruns -y status --seg_dbid 2
20170125:11:06:27:003020 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:27:003020 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-fault name:'execsort_mksort_mergeruns' fault type:'finish_pending' ddl statement:'' database name:'' table name:'' occurrence:'1' sleep time:'10' fault injection state:'completed'  num times hit:'1' 
20170125:11:06:27:003020 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- test if shared input scan deletes memory correctly when QueryFinishPending and its child has been eagerly freed,
-- where the child is a Sort node
drop table if exists testsisc;
NOTICE:  table "testsisc" does not exist, skipping
create table testsisc (i1 int, i2 int, i3 int, i4 int); 
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i1' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
insert into testsisc select i, i % 1000, i % 100000, i % 75 from
(select generate_series(1, nsegments * 50000) as i from 
	(select count(*) as nsegments from gp_segment_configuration where role='p' and content >= 0) foo) bar; 
set gp_resqueue_print_operator_memory_limits=on;
set statement_mem='2MB';
set gp_enable_mk_sort=off;
\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:27:003032 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:27:003032 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:27:003032 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- Set QueryFinishPending to true after SharedInputScan has retrieved the first tuple. 
-- This will eagerly free the memory context of shared input scan's child node.  
\! gpfaultinjector -f execshare_input_next -y finish_pending --seg_dbid 2
20170125:11:06:27:003047 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y finish_pending --seg_dbid 2
20170125:11:06:27:003047 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:27:003047 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
select COUNT(i2) over(partition by i1)
from testsisc
LIMIT 2;
 count 
-------
     1
     1
(2 rows)

\! gpfaultinjector -f execshare_input_next -y status --seg_dbid 2
20170125:11:06:28:003059 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y status --seg_dbid 2
20170125:11:06:28:003059 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:28:003059 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-fault name:'execshare_input_next' fault type:'finish_pending' ddl statement:'' database name:'' table name:'' occurrence:'1' sleep time:'10' fault injection state:'completed'  num times hit:'1' 
20170125:11:06:28:003059 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- test if shared input scan deletes memory correctly when QueryFinishPending and its child has been eagerly freed,
-- where the child is a Sort node and sort_mk algorithm is used
set gp_enable_mk_sort=on;
\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:28:003071 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:28:003071 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:28:003071 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
-- Set QueryFinishPending to true after SharedInputScan has retrieved the first tuple. 
-- This will eagerly free the memory context of shared input scan's child node.  
\! gpfaultinjector -f execshare_input_next -y finish_pending --seg_dbid 2
20170125:11:06:28:003083 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y finish_pending --seg_dbid 2
20170125:11:06:28:003083 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:28:003083 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
select COUNT(i2) over(partition by i1)
from testsisc
LIMIT 2;
 count 
-------
     1
     1
(2 rows)

\! gpfaultinjector -f execshare_input_next -y status --seg_dbid 2
20170125:11:06:28:003095 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y status --seg_dbid 2
20170125:11:06:28:003095 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:28:003095 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-fault name:'execshare_input_next' fault type:'finish_pending' ddl statement:'' database name:'' table name:'' occurrence:'1' sleep time:'10' fault injection state:'completed'  num times hit:'1' 
20170125:11:06:28:003095 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
reset gp_enable_mk_sort;
-- Disable faultinjectors
\! gpfaultinjector -f execsort_mksort_mergeruns -y reset --seg_dbid 2
20170125:11:06:29:003107 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execsort_mksort_mergeruns -y reset --seg_dbid 2
20170125:11:06:29:003107 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:29:003107 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
\! gpfaultinjector -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:29:003119 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Starting gpfaultinjector with args: -f execshare_input_next -y reset --seg_dbid 2
20170125:11:06:29:003119 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-Injecting fault on content=0:dbid=2:mode=s:status=u
20170125:11:06:29:003119 gpfaultinjector:nikos-mac:narmenatzoglou-[INFO]:-DONE
